### PR TITLE
Implement Queue#pop(non_block = false, timeout: nil)

### PR DESCRIPTION
**Pull request by casperisfine**
[Feature #18774]

I created `thread_sync.rb` for easier handling of keyword arguments.

### CODE REVIEW

File: test/ruby/test_settracefunc.rb [hunk 1]
```diff
@@ -2140,17 +2140,16 @@ def test_thread_add_trace_func
...
+    assert_equal ["line", base_line + 32],     events[0]
+    assert_equal ["line", base_line + 33],     events[1]
+    assert_equal ["call", base_line + -6],     events[2]
+    assert_equal ["return", base_line + -4],   events[3]
+    assert_equal ["line", base_line + 34],     events[4]
+    assert_equal ["line", base_line + 35],     events[5]
+    assert_equal ["c-call", base_line + 35],   events[6] # Thread.current
+    assert_equal ["c-return", base_line + 35], events[7] # Thread.current
+    assert_equal ["c-call", base_line + 35],   events[8] # Thread#set_trace_func
+    assert_equal nil,                          events[9]
```
Comment (ioquatix) [inline test/ruby/test_settracefunc.rb:2152]: Is this related to the PR?

File: thread.c [hunk 2]
```diff
@@ -1355,15 +1377,20 @@ rb_thread_sleep_interruptible(void)
 }
 
 static void
-rb_thread_sleep_deadly_allow_spurious_wakeup(VALUE blocker)
+rb_thread_sleep_deadly_allow_spurious_wakeup(VALUE blocker, VALUE timeout, rb_hrtime_t *end)
```
Comment (ioquatix) [inline thread.c:1380]: Are `VALUE timeout` and `rb_hrtime_t *end` the same things?

File: hrtime.h [hunk 3]
```diff
@@ -91,6 +92,22 @@ rb_hrtime_add(rb_hrtime_t a, rb_hrtime_t b)
     return c;
 }
 
+static inline rb_hrtime_t
+rb_hrtime_sub(rb_hrtime_t a, rb_hrtime_t b)
```
Comment (casperisfine) [inline hrtime.h:96]: I need `rb_hrtime_sub` to be able to compute the relative time from the absolute one. Since the sleep may be interrupted for other reasons and resumed, we pass around the final `end` time rather than a relative sleep time.

**Comment by ioquatix** [comment]
We need to investigate the way the fiber scheduler implementation handles timeouts, I suspect it will raise an exception, but it sounds like we should expand this definition to include "return nil" on timeout. Again, I suggested we have a consistent interface for this, e.g. `timeout: x, exception: true/false/nil`.

File: test/ruby/test_settracefunc.rb (hunk 1 above)
Comment (casperisfine) [reply test/ruby/test_settracefunc.rb:2152]: Yes, the move to `thread_sync.rb` made the `c-call` disapear.

File: thread.c (hunk 2 above)
Comment (casperisfine) [reply thread.c:1380]: Not quite. `timeout` is the relative duration e.g. `pop(timeout: 2)`.

I need it to pass it to the fiber scheduler if there is one.

If there is no fiber scheduler, I need the absolute timeout time, so that I can resume the sleep if we're interrupted for other reasons.

File: test/ruby/test_thread_queue.rb [hunk 4]
```diff
@@ -111,6 +111,15 @@ def test_queue_pop_interrupt
     assert_equal(0, q.num_waiting)
   end
 
+  def test_queue_pop_timeout
+    q = Thread::Queue.new
+    t1 = Thread.new { q.pop(timeout: 1) }
+    assert_nil t1.join(2).value
```
Comment (eregon) [inline test/ruby/test_thread_queue.rb:117]: This one test will take 1 second which is quite long, I'd use 0.1 or even better 0.001 to avoid wasting time during testing.

It would also be nice to test that it does return a pop'd value if there is one.

My bias: of course ruby/spec would be much better for this, and ensure other Ruby implementations have it sooner. For instance there is no way to run MRI tests of a more recent version, but that's supported by ruby/spec with PRETEND_RUBY_VERSION e.g. used in truffleruby, so e.g. it could be added already in TruffleRuby even though it's not aiming for full 3.2 compatibility yet.

**Review by eregon — APPROVED** [review]
Thank you for working on this, I was thinking I should try to make the time but I'm rather unfamiliar with the timeout stuff in CRuby.

File: test/ruby/test_thread_queue.rb (hunk 4 above)
Comment (casperisfine) [reply test/ruby/test_thread_queue.rb:117]: > This one test will take 1 second which is quite long, I'd use 0.1 or even better 0.001 to avoid wasting time during testing.

I agree, however the reason I test with an integer is that Integer and Float are two distinct codepaths.

> My bias: of course ruby/spec would be much better for this

Yeah, I'll add some specs. I'd like this to be reviewed by someone familiar with the thread scheduler first (likely @ko1), there's no point writing specs if my code is flawed.

Comment (eregon) [reply test/ruby/test_thread_queue.rb]: > there's no point writing specs if my code is flawed.

Just like tests it'd be extra confidence it works as expected, and it'd be useful regardless of the implementation :)

Comment (casperisfine) [reply test/ruby/test_thread_queue.rb]: I implemented `SizedQueue#push` and added some specs, let me know what you think.

**Comment by casperisfine** [comment]
I talked with @ko1, he approved the implementation with two small modifications, I'm now waiting on his confirmation that the implementation matches the spec agreed on by Matz.

**Review by ko1 — APPROVED** [review]

File: test/ruby/test_thread_queue.rb (hunk 4 above)
Comment (eregon) [reply test/ruby/test_thread_queue.rb]: > I [...] added some specs, let me know what you think.

Thanks! They look good.
